File: back-go/domain/statistics/besteffortcache_test.go
package statistics

import (
	"path/filepath"
	"testing"

	"mystravastats/domain/business"
	"mystravastats/domain/strava"
)

func TestBestEffortCache_SaveLoadRoundTrip(t *testing.T) {
	ClearBestEffortCache()
	stream := testBestEffortStream()
	supplierCalls := 0

	first := getOrComputeBestEffort(
		42,
		"best-time-distance",
		effortDistanceTarget(1000),
		stream,
		func() *business.ActivityEffort {
			supplierCalls++
			return &business.ActivityEffort{
				Distance: 1000,
				Seconds:  180,
				ActivityShort: business.ActivityShort{
					Id:   42,
					Name: "Warmup effort",
					Type: business.Ride,
				},
			}
		},
	)

	if supplierCalls != 1 {
		t.Fatalf("expected supplier to run once before persistence, got %d", supplierCalls)
	}
	if first == nil {
		t.Fatalf("expected first effort to be computed")
	}

	path := filepath.Join(t.TempDir(), "best-effort-cache.json")
	savedEntries, err := SaveBestEffortCacheToDisk(path)
	if err != nil {
		t.Fatalf("save cache failed: %v", err)
	}
	if savedEntries == 0 {
		t.Fatalf("expected at least one persisted cache entry")
	}

	ClearBestEffortCache()
	loadedEntries, err := LoadBestEffortCacheFromDisk(path)
	if err != nil {
		t.Fatalf("load cache failed: %v", err)
	}
	if loadedEntries != savedEntries {
		t.Fatalf("expected loaded entries %d, got %d", savedEntries, loadedEntries)
	}

	supplierCalls = 0
	second := getOrComputeBestEffort(
		42,
		"best-time-distance",
		effortDistanceTarget(1000),
		stream,
		func() *business.ActivityEffort {
			supplierCalls++
			return &business.ActivityEffort{
				Distance: 1000,
				Seconds:  999,
				ActivityShort: business.ActivityShort{
					Id:   42,
					Name: "Should not be used",
					Type: business.Ride,
				},
			}
		},
	)

	if supplierCalls != 0 {
		t.Fatalf("expected supplier not to run after cache reload, got %d calls", supplierCalls)
	}
	if second == nil || second.Seconds != 180 {
		t.Fatalf("expected cached effort to be reused with 180s, got %+v", second)
	}
}

func TestBestEffortCache_InvalidateByActivityIDs(t *testing.T) {
	ClearBestEffortCache()
	stream := testBestEffortStream()

	getOrComputeBestEffort(1, "best-time-distance", effortDistanceTarget(1000), stream, func() *business.ActivityEffort {
		return &business.ActivityEffort{Distance: 1000, Seconds: 200, ActivityShort: business.ActivityShort{Id: 1, Name: "A", Type: business.Ride}}
	})
	getOrComputeBestEffort(2, "best-time-distance", effortDistanceTarget(1000), stream, func() *business.ActivityEffort {
		return &business.ActivityEffort{Distance: 1000, Seconds: 210, ActivityShort: business.ActivityShort{Id: 2, Name: "B", Type: business.Ride}}
	})

	removed := InvalidateBestEffortCacheByActivityIDs(map[int64]struct{}{1: {}})
	if removed == 0 {
		t.Fatalf("expected invalidation to remove at least one entry")
	}
	if BestEffortCacheSize() == 0 {
		t.Fatalf("expected cache to retain entries for untouched activities")
	}
}

func testBestEffortStream() *strava.Stream {
	return &strava.Stream{
		Distance: strava.DistanceStream{Data: []float64{0, 500, 1000}, OriginalSize: 3},
		Time:     strava.TimeStream{Data: []int{0, 90, 180}, OriginalSize: 3},
		Altitude: &strava.AltitudeStream{Data: []float64{10, 20, 30}, OriginalSize: 3},
	}
}

File: back-kotlin/src/test/kotlin/me/nicolas/stravastats/domain/services/statistics/BestEffortCacheTest.kt
package me.nicolas.stravastats.domain.services.statistics

import me.nicolas.stravastats.domain.business.ActivityEffort
import me.nicolas.stravastats.domain.business.ActivityShort
import me.nicolas.stravastats.domain.business.ActivityType
import me.nicolas.stravastats.domain.business.strava.stream.AltitudeStream
import me.nicolas.stravastats.domain.business.strava.stream.DistanceStream
import me.nicolas.stravastats.domain.business.strava.stream.Stream
import me.nicolas.stravastats.domain.business.strava.stream.TimeStream
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Path

class BestEffortCacheTest {

    @TempDir
    lateinit var tempDir: Path

    @Test
    fun `best effort cache is persisted and reloaded from disk`() {
        // GIVEN
        BestEffortCache.clear()
        val stream = testStream()
        var calls = 0
        val effort = ActivityEffort(
            distance = 1000.0,
            seconds = 180,
            deltaAltitude = 20.0,
            idxStart = 0,
            idxEnd = 2,
            averagePower = 250,
            label = "Best speed for 1000m",
            activityShort = ActivityShort(42, "Warmup effort", ActivityType.Ride.name),
        )

        // WHEN
        val first = BestEffortCache.getOrCompute(
            activityId = 42,
            metric = "best-time-distance",
            target = "1000.0",
            stream = stream,
        ) {
            calls += 1
            effort
        }

        // THEN
        assertEquals(1, calls)
        assertNotNull(first)

        // WHEN - persisted and reloaded
        val cacheFile = tempDir.resolve("best-effort-cache.json")
        val persisted = BestEffortCache.saveToDisk(cacheFile)
        assertEquals(1, persisted)

        BestEffortCache.clear()
        val loaded = BestEffortCache.loadFromDisk(cacheFile)
        assertEquals(persisted, loaded)

        calls = 0
        val second = BestEffortCache.getOrCompute(
            activityId = 42,
            metric = "best-time-distance",
            target = "1000.0",
            stream = stream,
        ) {
            calls += 1
            null
        }

        // THEN - loaded from disk, no recomputation
        assertEquals(0, calls)
        assertEquals(180, second?.seconds)
    }

    @Test
    fun `invalidateActivities removes only targeted entries`() {
        // GIVEN
        BestEffortCache.clear()
        val stream = testStream()

        BestEffortCache.getOrCompute(1, "best-time-distance", "1000.0", stream) {
            ActivityEffort(
                distance = 1000.0,
                seconds = 200,
                deltaAltitude = 10.0,
                idxStart = 0,
                idxEnd = 2,
                averagePower = null,
                label = "A",
                activityShort = ActivityShort(1, "A", ActivityType.Ride.name),
            )
        }
        BestEffortCache.getOrCompute(2, "best-time-distance", "1000.0", stream) {
            ActivityEffort(
                distance = 1000.0,
                seconds = 210,
                deltaAltitude = 11.0,
                idxStart = 0,
                idxEnd = 2,
                averagePower = null,
                label = "B",
                activityShort = ActivityShort(2, "B", ActivityType.Ride.name),
            )
        }

        // WHEN
        val removed = BestEffortCache.invalidateActivities(setOf(1L))

        // THEN
        assertEquals(1, removed)
        assertEquals(1, BestEffortCache.size())

        val missing = BestEffortCache.getOrCompute(1, "best-time-distance", "1000.0", stream) { null }
        assertNull(missing)
    }

    private fun testStream(): Stream = Stream(
        distance = DistanceStream(
            data = listOf(0.0, 500.0, 1000.0),
            originalSize = 3,
            resolution = "high",
            seriesType = "distance",
        ),
        time = TimeStream(
            data = listOf(0, 90, 180),
            originalSize = 3,
            resolution = "high",
            seriesType = "time",
        ),
        altitude = AltitudeStream(
            data = listOf(100.0, 110.0, 120.0),
            originalSize = 3,
            resolution = "high",
            seriesType = "distance",
        ),
    )
}
